exchanges/poloniex/poloniex_websocket.go

```diff
@@ -626,51 +588,48 @@ func (p *Poloniex) Unsubscribe(unsub []subscription.Subscription) error {
 			return err
 		}
 	}
+
 	var errs error
-channels:
-	for i := range unsub {
-		unsubscriptionRequest := WsCommand{
-			Command: "unsubscribe",
+	for _, s := range subs {
+		var err error
+		if creds != nil && strings.EqualFold(strconv.FormatInt(wsAccountNotificationID, 10), s.Channel) {
+			err = p.wsSendAuthorisedCommand(creds.Secret, creds.Key, op)
+		} else {
+			req := wsCommand{Command: op}
+			if strings.EqualFold(strconv.FormatInt(wsTickerDataID, 10), s.Channel) {
+				req.Channel = wsTickerDataID
+			} else {
+				if len(s.Pairs) != 1 {
+					return subscription.ErrNotSinglePair
+				}
+				req.Channel = s.Pairs[0].String()
+			}
+			err = p.Websocket.Conn.SendJSONMessage(req)
 		}
-		switch {
-		case strings.EqualFold(strconv.FormatInt(wsAccountNotificationID, 10),
-			unsub[i].Channel) && creds != nil:
-			err := p.wsSendAuthorisedCommand(creds.Secret, creds.Key, "unsubscribe")
-			if err != nil {
-				errs = common.AppendError(errs, err)
-				continue channels
+		if err == nil {
+			if op == wsSubscribeOp {
+				err = p.Websocket.AddSuccessfulSubscriptions(s)
+			} else {
+				err = p.Websocket.RemoveSubscriptions(s)
 			}
-			p.Websocket.RemoveSubscriptions(unsub[i])
-			continue channels
-		case strings.EqualFold(strconv.FormatInt(wsTickerDataID, 10),
-			unsub[i].Channel):
-			unsubscriptionRequest.Channel = wsTickerDataID
-		default:
-			unsubscriptionRequest.Channel = unsub[i].Pair.String()
 		}
-		err := p.Websocket.Conn.SendJSONMessage(unsubscriptionRequest)
 		if err != nil {
```

Comment:
This can be shifted above the `err == nil` check and then you can rm the `err == nil` statement

Reply:
The alternative I see is:
```
			err = p.Websocket.Conn.SendJSONMessage(req)
		}
		if err != nil {
			errs = common.AppendError(errs, err)
			continue
		}
		if op == wsSubscribeOp {
			err = p.Websocket.AddSuccessfulSubscriptions(s)
		} else {
			err = p.Websocket.RemoveSubscriptions(s)
		}
		if err != nil {
			errs = common.AppendError(errs, err)
		}
```

I chose the pattern I have deliberately. But this new pattern of `if err == nil {` is new in our codebase.
If the action on err was heavier I might wrap it in a anonyfunc:

```	for _, s := range subs {
		if err := func() {
			var err error
			if creds != nil && strings.EqualFold(strconv.FormatInt(wsAccountNotificationID, 10), s.Channel) {
				err = p.wsSendAuthorisedCommand(creds.Secret, creds.Key, op)
			} else {
				req := wsCommand{Command: op}
				if strings.EqualFold(strconv.FormatInt(wsTickerDataID, 10), s.Channel) {
					req.Channel = wsTickerDataID
				} else {
					if len(s.Pairs) != 1 {
						return subscription.ErrNotSinglePair
					}
					req.Channel = s.Pairs[0].String()
				}
				err = p.Websocket.Conn.SendJSONMessage(req)
			}
			if err != nil {
				return err
			}
			if op == wsSubscribeOp {
				return p.Websocket.AddSuccessfulSubscriptions(s)
			}
			return p.Websocket.RemoveSubscriptions(s)
		}(); err != nil {
			errs = common.AppendError(errs, err)
		}
	}
```
but that doesn't feel justified in this case.

Reply:
Ahh yeah I see it now, only thing I can suggest  is I would return fatal and just collect non-fatal, but I would even go so far as making it more strict and return the error on removing and adding subs, I am all for that strict life. 

```go
	var errs error
	for _, s := range subs {
		var err error
		if creds != nil && strings.EqualFold(strconv.FormatInt(wsAccountNotificationID, 10), s.Channel) {
			err = p.wsSendAuthorisedCommand(creds.Secret, creds.Key, op)
		} else {
			req := wsCommand{Command: op}
			if strings.EqualFold(strconv.FormatInt(wsTickerDataID, 10), s.Channel) {
				req.Channel = wsTickerDataID
			} else {
				if len(s.Pairs) != 1 {
					return subscription.ErrNotSinglePair
				}
				req.Channel = s.Pairs[0].String()
			}
			err = p.Websocket.Conn.SendJSONMessage(req)
		}
		if err != nil {
			return err
		}

		if op == wsSubscribeOp {
			errs = common.AppendError(errs, p.Websocket.AddSuccessfulSubscriptions(s))
		} else {
			errs = common.AppendError(errs, p.Websocket.RemoveSubscriptions(s))
		}
	}
	return errs
```

Reply:
Strict is okay, but so is informative.
Definitely don't want to bail on the first sub error from SendJSONMessage.

Tempted by the simplification of calling AppendError on Add/Remove, but I think gloriousCode will justifiably note that it hides the informative part of the line too much, and implies that it *is* an error, not *might be*.

Imma leave as is, since I think we're pretty subjective throughout here.